### ISSUE TITLE
sql: Cannot set empty STRING[] and INT[] columns in same UPDATE query

### ISSUE THREAD

**Issue by slugalisk** [cockroachdb/cockroach#23429]
**Log headers:**
```
I180305 15:28:38.078461 78 util/log/clog.go:889  [config] file created at: 2018/03/05 15:28:38
I180305 15:28:38.078461 78 util/log/clog.go:889  [config] running on machine: cockroachdb-0
```
```
I180305 18:25:12.512595 78 util/log/clog.go:889  [config] file created at: 2018/03/05 18:25:12
I180305 18:25:12.512595 78 util/log/clog.go:889  [config] running on machine: cockroachdb-1
I180305 18:25:12.512595 78 util/log/clog.go:889  [config] binary: CockroachDB CCL v1.1.3 (linux amd64, built 2017/11/27 13:59:10, go1.8.3)
I180305 18:25:12.512595 78 util/log/clog.go:889  [config] arguments: [/cockroach/cockroach start --logtostderr --insecure --host cockroachdb-1.cockroachdb.test.svc.cluster.local --http-host 0.0.0.0 --cache 25% --max-sql-memory 25% --join cockroachdb-public]
```
```
I180305 02:20:52.324948 104 util/log/clog.go:889  [config] file created at: 2018/03/05 02:20:52
I180305 02:20:52.324948 104 util/log/clog.go:889  [config] running on machine: cockroachdb-2
I180305 02:20:52.324948 104 util/log/clog.go:889  [config] binary: CockroachDB CCL v1.1.3 (linux amd64, built 2017/11/27 13:59:10, go1.8.3)
I180305 02:20:52.324948 104 util/log/clog.go:889  [config] arguments: [/cockroach/cockroach start --logtostderr --insecure --host cockroachdb-2.cockroachdb.test.svc.cluster.local --http-host 0.0.0.0 --cache 25% --max-sql-memory 25% --join cockroachdb-public]
```

**Issue:**
An error is returned when attempting to set both a `STRING[]` and `INT[]` column to an empty array in the same query. Adding type hints does not fix the problem...

**Steps to reproduce:**
```
> CREATE TABLE test (id SERIAL, str_array STRING[], int_array INT[]);
CREATE TABLE

Time: 225.563381ms

> INSERT INTO test (str_array, int_array) VALUES (ARRAY['a', 'b', 'c'], ARRAY[1, 2, 3]) RETURNING id;
+--------------------+
|         id         |
+--------------------+
| 328385406202478597 |
+--------------------+
(1 row)

Time: 1.565237587s

> UPDATE test SET str_array = ARRAY[], int_array = ARRAY[] WHERE id = 328385406202478597;
pq: value type string[] doesn't match type ARRAY of column "int_array"
> UPDATE test SET str_array = ARRAY[]:::string[], int_array = ARRAY[]:::int[] WHERE id = 328385406202478597;
pq: value type string[] doesn't match type ARRAY of column "int_array"
```

FWIW these all work... the bug only occurs during `UPDATE` queries and only when both values are empty.
```
> UPDATE test SET str_array = ARRAY[] WHERE id = 328385406202478597;
UPDATE 1

Time: 49.112494ms

> UPDATE test SET int_array = ARRAY[] WHERE id = 328385406202478597;
UPDATE 1

Time: 17.422961ms

> UPDATE test SET str_array = ARRAY['a'] WHERE id = 328385406202478597;
UPDATE 1

Time: 460.107682ms

> UPDATE test SET int_array = ARRAY[1] WHERE id = 328385406202478597;
UPDATE 1

Time: 16.568802ms

> UPDATE test SET str_array = ARRAY['a'], int_array = ARRAY[1] WHERE id = 328385406202478597;
UPDATE 1

Time: 15.254256ms

> UPDATE test SET str_array = ARRAY[], int_array = ARRAY[1] WHERE id = 328385406202478597;
UPDATE 1

Time: 314.671008ms

> UPDATE test SET str_array = ARRAY['a'], int_array = ARRAY[] WHERE id = 328385406202478597;
UPDATE 1

Time: 1.778539935s

> INSERT INTO test (str_array, int_array) VALUES (ARRAY[], ARRAY[]) RETURNING id;
+--------------------+
|         id         |
+--------------------+
| 328390044028207109 |
+--------------------+
(1 row)

Time: 2.668086147s
```

**Comment by dt** [comment]
@justinj @jordanlewis IIRC y'all are the arrays experts or at least would know who is and can re-assign, or if this is just a typing rather than array-speicifc issue.

**Comment by justinj** [comment]
@slugalisk This will be fixed in the next release via #23465 - thanks so much, the detailed bug report is very much appreciated!

**Comment by slugalisk** [comment]
@justinj awesome, thank you for resolving this so quickly!
